medcat/utils/saving/envsnapshot.py

```diff
@@ -0,0 +1,73 @@
+from typing import List, Dict, Any, Set
+
+import os
+import re
+import pkg_resources
+import platform
+
+
+ENV_SNAPSHOT_FILE_NAME = "environment_snapshot.json"
+
+INSTALL_REQUIRES_FILE_PATH = os.path.join(os.path.dirname(__file__),
+                                          "..", "..", "..",
+                                          "install_requires.txt")
+# NOTE: The install_requires.txt file is copied into the wheel during build
+#       so that it can be included in the distributed package.
+#       However, that means it's 1 folder closer to this file since it'll now
+#       be in the root of the package rather than the root of the project.
+INSTALL_REQUIRES_FILE_PATH_PIP = os.path.join(os.path.dirname(__file__),
+                                              "..", "..",
+                                              "install_requires.txt")
+
+
+def get_direct_dependencies() -> Set[str]:
+    """Get the set of direct dependeny names.
+
+    The current implementation reads install_requires.txt for dependenceies,
+    removes comments, whitespace, quotes; removes the versions and returns
+    the names as a set.
+
+    Returns:
+        Set[str]: The set of direct dependeny names.
+    """
+    req_file = INSTALL_REQUIRES_FILE_PATH
+    if not os.path.exists(req_file):
+        # When pip-installed. See note above near constant definiation
+        req_file = INSTALL_REQUIRES_FILE_PATH_PIP
+    with open(req_file) as f:
+        # read every line, strip quotes and comments
+        dep_lines = [line.split("#")[0].replace("'", "").replace('"', "").strip() for line in f.readlines()]
+        # remove comment-only (or empty) lines
+        deps = [dep for dep in dep_lines if dep]
+    return set(re.split("[@<=>~]", dep)[0].strip() for dep in deps)
+
+
+def get_installed_packages() -> List[List[str]]:
+    """Get the installed packages and their versions.
+
+    Returns:
+        List[List[str]]: List of lists. Each item contains of a dependency name and version.
+    """
+    direct_deps = get_direct_dependencies()
+    installed_packages = []
+    for package in pkg_resources.working_set:
+        if package.project_name not in direct_deps:
+            continue
```

Comment:
IMHO why not save transitive dependencies in the snapshot as well, together with direct dependencies? It would be easier for users like me to create a virtual environment with high fidelity to the original training environment before starting to use the model.

Reply:
There's a few issues with listing all installed packages:
- The list of dependencies may change depending on hardware (i.e nvidia or cuda related packages)
- The version of dependencies may change depending for different python versions
- Some environments may have unrelated packages

With that said, that was my original idea as well. If we're able to identify the exact environment (down to each individual package) then we _should_ be able to **exactly** replicate behaviour across environments.

But the main reason to go with the current approach is the fact that downstream / transitive dependencies aren't (or at least shouldn't be) our responsibility. Our direct dependencies _should_ know what they can and cannot work with.
There are, of course, situations where that expectation fails. Especially since we depend on quite a few older versions of other projects.

So all in all, I think it would be worth it to add something to this in the future (as part of another PR). Whether it's all installed packages or some subset of it would need to be figured out as part of that.
But I think this PR will still allow as to get at least 90% of the way there in terms of being able to validate our environment.